File: algos_semina/5th_seminar/seminar_2/3rd_grade/이서진_metamong/20438.py
import sys
n,k,q,m=map(int,sys.stdin.readline().split())
sleep=[0 for _ in range(n+3)]
chk=[0 for _ in range(n+3)]
psum=[0 for _ in range(n+3)]
for _ in range(k):
    t=int(sys.stdin.readline())
    sleep[t]=1
x=list(map(int,sys.stdin.readline().split())) 
for i in x:
    if sleep[i]==1 :
        continue
    hi=i
    while i<=n+2:
        if sleep[i]==1:
            i+=hi
            continue
        chk[i]=1
        i+=hi

for i in range(3,n+3):
    psum[i]=psum[i-1]+(chk[i]==0)
for i in range(0,m):
    s,e=map(int,sys.stdin.readline().split())
    print(psum[e]-psum[s-1])
    print()

File: algos_semina/5th_seminar/seminar_2/3rd_grade/심채영_visionWWW/20438.cpp
#include <iostream>
#include <vector>
using namespace std;

int n, k, q, m, x, s, e, add;
bool stdarr[5005], sleep[5005], chk[5005];
 
int main() {

	cin >> n >> k >> q >> m;

	while (k--) {
		cin >> x;
		sleep[x] = 1;
	}

	while (q--) {
		cin >> x;
		if (sleep[x]) continue;

		add = x;

		while (x <= n + 2) {
			if (sleep[x]) {
				x += add;
				continue;
			}
			chk[x] = 1;
      x += add;
		}
	}
	for (int i = 3; i <= n + 2; i++) stdarr[i] = stdarr[i - 1] + (chk[i] == 0);
	for (int i = 0; i < m; i++) {
		cin >> s >> e;
		cout << stdarr[e] - stdarr[s - 1] << '\n';
	}
}
